### ISSUE TITLE
Improve documentation for onPanDown/onPanStart

### ISSUE THREAD

**Issue by stuz5000** [flutter/flutter#50776]
With a tree of widgets:
``` 
 - GestureDetector (onPanDown)
   - Container (green)
     -  GestureDetector (onPanDown, HitTestBehavior.opaque)
        - Container (red)
```
... both get the event if the inner (opaque red box is tapped). If I switch to:
``` 
 - Listener (onPointerDown)
   - Container (green)
     -  GestureDetector (onPanDown, HitTestBehavior.opaque)
        - Container (red)
```
... again, the event is received twice for one tap.

For both, because `behavior:HitTestBehavior.opaque` is set for the inner box, I don't expect the event to be received twice.

Happens on both web and mobile emulator.

---
Why not use onTapDown? Suppose I wanted to:
 
 - eagerly detect double taps in the green box AND drag events in the green box,
 - UNLESS a gesture started in an opaque child, such as a button. 

(I can't use two GestureDetectors here because it waits 500ms until the completion of a gesture before issuing the tapdown/pandown events, which feels like UI jank to the user.)

```dart
import 'package:flutter/material.dart';

void main() => runApp(MaterialApp(
  home: Scaffold(body: Center(child: Content())),
));

class Content extends StatelessWidget {
  Content({Key key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    Widget child = Container(
        width: 50,
        height: 50,
        child: Center(child: Text("Tap me")),
        color: Colors.blue);
    child = GestureDetector(
      // *should* prevent redbox receiving events.
      // This is only true if redbox is wrapped with GestureDetector, not
      // Listener.
      behavior: HitTestBehavior.opaque,
//    behavior: HitTestBehavior.deferToChild,
//    behavior: HitTestBehavior.translucent,

//      onTap: () {
//        print("onTap blue");
//      },
//    onTapDown: (_) {
//      print("onTapDown blue");
//    },
    onPanDown: (_) {
      print("onPanDown blue");
    },
      child: child,
    );

//    // IgnorePointer / absorb pointer don't really help
//    // If uncommented, only the parent (red) box gets events
//   child = IgnorePointer(child:child);
//    // If uncommented, only the parent (red) box gets events
//   child = AbsorbPointer(child:child);

    child = Center(child: child);
    child = Container(color: Colors.red, child: child);

    // Listener also gets the same event as child...
    child = Listener(
// No adjustments here help
//    behavior: HitTestBehavior.deferToChild,  // default
//    behavior: HitTestBehavior.opaque,
//    behavior: HitTestBehavior.translucent,

      onPointerDown: (details) {
        print("onPointerDown red");
      },
      child: child,
    );

//    // Does nested GestureDetector work? Kinda.
//    // This *does* correctly get blocked from receiving
//    // events when an opaque child is hit, this and blue's
//    // both listen to onTap, or
//    // both listen to onTapDown.
//    // If both listen to onPanDown,
//    // both get events event though blue is opaque.
//  child = GestureDetector(
////    behavior: HitTestBehavior.deferToChild,
//    behavior: HitTestBehavior.opaque,
////    behavior: HitTestBehavior.translucent,
//
//    // Uncommenting this means this detector receives no
//    // events when its child is tapped. Oddly its the
//    // onTapDown, not the onTap callback that gets fired.
////    onTap: (){
////      print("onTap red");
////    },
////    onTapDown: (details) {
////      print("onTapDown red");
////    },
//    onPanDown: (_) {
//      print("onPanDown red");
//    },
//    child:child,
//  );
    child = SizedBox(width: 300, height: 300, child: child);
    return child;
  }
}
```
```
flutter doctor -v
[✓] Flutter (Channel master, v1.15.4-pre.60, on Linux, locale en_US.UTF-8)
    • Flutter version 1.15.4-pre.60 at /home/sir/development/flutter
    • Framework revision e2554a9202 (7 hours ago), 2020-02-13 11:56:45 -0800
    • Engine revision 0e8ed2715a
    • Dart version 2.8.0 (build 2.8.0-dev.8.0 5244d99a5d)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /home/sir/Android/Sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • ANDROID_HOME = /home/sir/Android/Sdk
    • Java binary at: /home/sir/soft/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Android Studio (version 3.5)
    • Android Studio at /home/sir/soft/android-studio
    • Flutter plugin version 43.0.1
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] IntelliJ IDEA Ultimate Edition (version 2019.3)
    • IntelliJ at /snap/intellij-idea-ultimate/204
    • Flutter plugin version 42.1.4
    • Dart plugin version 193.5731

[✓] Connected device (3 available)
    • Android SDK built for x86 • emulator-5554 • android-x86    • Android 10 (API 29) (emulator)
    • Chrome                    • chrome        • web-javascript • Google Chrome 76.0.3809.132
    • Web Server                • web-server    • web-javascript • Flutter Tools

• No issues found!
```

Related to discussion here: https://github.com/flutter/flutter/issues/18450

**Comment by rydmike** [comment]
I have the same issue and question, here shown with a similar but maybe a bit simpler formatted example:

```
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Gesture Issue',
      home: Scaffold(body: Center(child: GestureIssue())),
    );
  }
}

class GestureIssue extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onPanDown: (_) {
        print('onPanDown pink!');
      },
      onTap: () {
        print("onTap pink!");
      },
      child: Container(
        width: 250,
        height: 250,
        color: Colors.pink[100],
        child: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.spaceEvenly,
            children: <Widget>[
              Text('Tap me!'),
              GestureDetector(
                behavior: HitTestBehavior.opaque,
                onTap: () {
                  print("onTap blue!");
                },
                onPanDown: (_) {
                  print("onPanDown blue!");
                },
                child: Container(
                  width: 100,
                  height: 100,
                  color: Colors.blue[300],
                  child: Center(child: Text('Tap me!')),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

```

How do we prevent pan events from also being received and detected by the "Pink" container?

**Use case scenario (my actual case):** 
Consider a packaged custom widget that uses onPanStart, OnPanDown, OnPanUpdate for interaction. If such a widget is on a scrolling surface, how do we block the pan events from being forwarded from such a widget to whatever parent is in charge of detecting scrolling events?

When the user interacts with such a custom widget with its own gesture detection handling, it most likely needs block all gestures to parent(s) and hence also scroll events to parent(s). The actual widget in my use case works fine as long as it is not on a scrolling surface. It also works on web/desktop as soon as the surface is large enough that there is no need for scrolling, even if it scrolls when it is smaller, eg if everything is wrapped in a singleChildScrollView. In such a case as soon as the widget is on a scrolling surface it becomes impossible/difficult to interact with. If the scrolling of the parent is in the vertical direction, you can get around the issue by panning horizontally to catch the events only in the  widget in question, but there is constantly ambiguous events fired to the parent scroll controller. 

 Imo the:

behavior: HitTestBehavior.opaque,

on the child does not work as expected or I have just misunderstood how it is supposed to work. Maybe there is some other way to accomplish the mentioned task in question, but I have not found it yet, despite researching the topic quite a bit.

I did my test on master, v1.18.1-pre.11 too, it can be seen there too, but the result is the same on current online version of Dartpad.

**Comment by rydmike** [comment]
Here is the above example as a Gist in a DartPad:
https://dartpad.dartlang.org/ea5c2d38d03642bf1ad9cf70893828e0

The issue or question if it can be solved in another way, how can the "blue" container and its GestureDetector prevent touch events on itself, to propagate to its parent, without knowing anything about its parent(s)? 

I kind of assumed using "**behavior: HitTestBehavior.opaque**" would be the answer, but apparently it is not. Maybe there is another way that is just not obvious to me?

**Comment by rydmike** [comment]
**TLDR**

GestureDetector `behavior: HitTestBehavior.opaque` does **not** seem to do what is claimed in the API docs: https://api.flutter.dev/flutter/rendering/HitTestBehavior-class.html

Either it is broken or the documented explanation of how it is supposed to work or what it actually does, needs improvements.

@TahaTesser do you have any knowledge or input on the issue?
@stuz5000 the workaround I present further below worked for my use case. It might or might not be applicable to your actual use case too, but check it out if this case is still an issue for you.

----

**Here is another presentation of the onPan issue:**

This example shows the issue with using onPan events with gesture detector. The issue is especially obvious when the gesture detector is on a scrolling surface that intercepts the scroll events before the pan events and consumes them.

When considering this problem we should also specify that an acceptable solution can only change and adjust code withing the blue widget with its gesture detector. The solution may not modify scroll event handler as the solution needs to work also when the "blue" widget is a package widget that is unaware of the circumstances it is used in and it cannot be expected that consumer of the Widget should have to modify other properties than the Widget constructor.

See and try the issue in:
- DartPad: https://dartpad.dartlang.org/adc00ee96cf8cf454790e10c068e2ccc
- CodePen: https://codepen.io/rydmike/pen/gOaQvrv

The gist for it is here: https://gist.github.com/rydmike/adc00ee96cf8cf454790e10c068e2ccc

Why is `behavior: HitTestBehavior.opaque` not working as expected and blocking events from reaching the scroll handler? Also this example uses a more complex SliverList for the scrolling. This actually behaves even more worse than when the custom Widget with its gesture detector is in SingleChildScrollView.

![image](https://user-images.githubusercontent.com/39990307/82268886-c34dc300-9978-11ea-83c4-be06bb4092e8.png)

--------

**Possible workaround**

Instead of using onPan... event handlers I tried working around the issue by using onHorizontalDrag and onVerticalDrag event handlers.  Doing so actually works, and it works very well. 

I don't think the `behavior: HitTestBehavior.opaque` has any impact here either. I suspect the lower level scroll handler uses onPan (or equivalent that consumes onPans) and for some reason as it is not using the drag events, the custom widget can use those events as they are not intercepted by event handler when scrolling. Well so it seems, at least it works the way I need it to with a bunch of onHorizontalDrag and onVerticalDrag handlers instead of onPan variants.

See and try the workaround I used here in:
- DartPad: https://dartpad.dartlang.org/adb2de511815c300b66ad0f4ca76bb63
- CodePen: https://codepen.io/rydmike/pen/RwWqQGL

The gist for it is here: https://gist.github.com/rydmike/adb2de511815c300b66ad0f4ca76bb63

![image](https://user-images.githubusercontent.com/39990307/82268855-a4e7c780-9978-11ea-9d51-741fdf39a1aa.png)

--------

**Comment by TahaTesser** [comment]
Using []code sample(https://github.com/flutter/flutter/issues/50776#issuecomment-610939342)

### When tapped in pink area
```
D/ViewRootImpl@dd4866e[MainActivity](31927): ViewPostIme pointer 0
I/flutter (31927): onPanDown pink!
D/ViewRootImpl@dd4866e[MainActivity](31927): ViewPostIme pointer 1
I/flutter (31927): onTap pink!

```
### When tapped in blue area
```
D/ViewRootImpl@dd4866e[MainActivity](31927): ViewPostIme pointer 0
I/flutter (31927): onPanDown blue!
I/flutter (31927): onPanDown pink!
D/ViewRootImpl@dd4866e[MainActivity](31927): ViewPostIme pointer 1
I/flutter (31927): onTap blue!

```
<details> 
<summary>flutter doctor -v</summary> 

``` 
[✓] Flutter (Channel dev, 1.19.0-1.0.pre, on Mac OS X 10.15.4 19E287, locale
    en-GB)
    • Flutter version 1.19.0-1.0.pre at /Users/tahatesser/Code/flutter_dev
    • Framework revision 456d80b9dd (8 days ago), 2020-05-11 11:45:03 -0400
    • Engine revision d96f962ca2
    • Dart version 2.9.0 (build 2.9.0-7.0.dev 092ed38a87)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/tahatesser/Code/SDK
    • Platform android-29, build-tools 29.0.3
    • ANDROID_HOME = /Users/tahatesser/Code/SDK
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_212-release-1586-b4-5784211)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.4.1, Build version 11E503a
    • CocoaPods version 1.9.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.6)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 45.1.1
    • Dart plugin version 192.8052
    • Java version OpenJDK Runtime Environment (build
      1.8.0_212-release-1586-b4-5784211)

[✓] VS Code (version 1.45.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.10.2

[✓] Connected device (4 available)
    • Taha’s iPhone • 00008020-001059882212002E • ios            • iOS 13.4.1
    • macOS         • macOS                     • darwin-x64     • Mac OS X
      10.15.4 19E287
    • Web Server    • web-server                • web-javascript • Flutter Tools
    • Chrome        • chrome                    • web-javascript • Google Chrome
      81.0.4044.138

• No issues found!
``` 

</details>

**Comment by rydmike** [comment]
Thanks @TahaTesser! 

To me the question is why does onPanDown events pass to pink when you onPanDown on blue, even if blue has `HitTestBehavior.opaque` set? Should it not then block that from happening as it says in: https://api.flutter.dev/flutter/rendering/HitTestBehavior-class.html ?

It does not appear to do so. This has many consequences, if e.g. then underlying widgets are basing their functionality on onPan (or maybe some lower level equivalent of it), like maybe scrolling vertically probably does, it means that you cannot implement widgets on top of such a surface that for their own hit area need to block the events from being consumed by the widget below it in order to work properly.

If this is really the case, then to me it sounds like this issue: https://github.com/flutter/flutter/issues/35452 might be a victim of this issue as well.

To some extent it is possible to work around this issue, like demonstrated in my more elaborate example above.

**Comment by rydmike** [comment]
@mariamhas Just wanted to let you know that this is an issue I have noticed as well, even if I did not open the original issue. **It is potentially a real issue in all versions of Flutter.** 

Why does onPanDown events pass to pink when you onPanDown on blue, even if blue has `HitTestBehavior.opaque` set? Should it not then block that from happening, as is stated in: https://api.flutter.dev/flutter/rendering/HitTestBehavior-class.html ?

Potentially it might also be an issue where we do not understand the documentation the way it is written now, or it says that it does more than it actually does and should say something else, or potentially `HitTestBehavior.opaque` is not working as intended, in that case it is a **bug**.

For my own use case I found and documented a usable workaround above, other use cases might not be able to use it, but it should work in many cases.

**Comment by darshankawar** [comment]
<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 1.22.6, on Mac OS X 10.15.4 19E2269 darwin-x64,
    locale en-IN)
    • Flutter version 1.22.6 at /Users/dhs/documents/Fluttersdk/flutter
    • Framework revision 9b2d32b605 (6 days ago), 2021-01-22 14:36:39 -0800
    • Engine revision 2f0af37152
    • Dart version 2.10.5

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at /Users/dhs/Library/Android/sdk
    • Platform android-30, build-tools 30.0.0
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    • CocoaPods version 1.9.3

[!] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.52.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.16.0

[✓] Connected device (2 available)
    • SM A260G (mobile)            • 5200763ebcfa861f • android-arm • Android
      8.1.0 (API 27)
    • AOSP on IA Emulator (mobile) • emulator-5554    • android-x86 • Android 9
      (API 28) (emulator)

! Doctor found issues in 1 category.

[✓] Flutter (Channel master, 1.26.0-13.0.pre.229, on Mac OS X 10.15.4 19E2269
    darwin-x64, locale en-IN)
    • Flutter version 1.26.0-13.0.pre.229 at
      /Users/dhs/documents/Fluttersdk/flutter
    • Framework revision 650b240634 (5 hours ago), 2021-01-27 21:39:06 -0800
    • Engine revision 68cd4c42b5
    • Dart version 2.12.0 (build 2.12.0-267.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at /Users/dhs/Library/Android/sdk
    • Platform android-30, build-tools 30.0.0
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[!] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    ! CocoaPods 1.9.3 out of date (1.10.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin
        code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see
      https://guides.cocoapods.org/using/getting-started.html#installation for
      instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.52.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.16.0

[✓] Connected device (4 available)
    • SM A260G (mobile)            • 5200763ebcfa861f • android-arm    • Android
      8.1.0 (API 27)
    • AOSP on IA Emulator (mobile) • emulator-5554    • android-x86    • Android
      9 (API 28) (emulator)
    • macOS (desktop)              • macos            • darwin-x64     • Mac OS
      X 10.15.4 19E2269 darwin-x64
    • Chrome (web)                 • chrome           • web-javascript • Google
      Chrome 88.0.4324.96

! Doctor found issues in 1 category.


```

</details>

```
I/flutter (11963): onPanDown blue!
I/flutter (11963): onPanDown pink!
I/flutter (11963): onTap blue!
I/flutter (11963): onPanDown pink!
I/flutter (11963): onTap pink!
I/flutter (11963): onPanDown blue!
I/flutter (11963): onPanDown pink!
I/flutter (11963): onTap blue!
I/flutter (11963): onPanDown blue!
I/flutter (11963): onPanDown pink!
I/flutter (11963): onTap blue!
```

**Comment by stuz5000** [comment]
Any progress here -- is this a bug?
Is there any workaround to block onPanDown events from being passed to parent widgets?

**Comment by Hixie** [comment]
This is working as intended (see https://github.com/flutter/flutter/issues/74733#issuecomment-767859584).

onPanDown is intentionally sent to anywhere that _might_ win the gesture arena. It does not indicate a pan has started. It's like onTapDown.

onPanStart is only called for the one gesture detector that won the arena.

**Comment by goderbauer** [comment]
Closing this as per the previous comment.
